core/executor/src/wasm_executor.rs

```diff
@@ -121,1201 +137,1136 @@ fn deadline_to_timestamp(deadline: u64) -> Option<offchain::Timestamp> {
-			.map_err(|_| "Invalid attempt to determine key in ext_set_storage")?;
-		let value = this.memory.get(value_data, value_len as usize)
-			.map_err(|_| "Invalid attempt to determine value in ext_set_storage")?;
-		if let Some(_preimage) = this.hash_lookup.get(&key) {
-			debug_trace!(
-				target: "wasm-trace",
-				"*** Setting storage: %{} -> {}   [k={}]",
-				primitives::hexdisplay::ascii_format(&_preimage),
-				HexDisplay::from(&value),
-				HexDisplay::from(&key),
-			);
+		&Resolver
+	}
+}
+
+impl wasmi::Externals for FunctionExecutor {
+	fn invoke_index(&mut self, index: usize, args: wasmi::RuntimeArgs)
+		-> std::result::Result<Option<wasmi::RuntimeValue>, wasmi::Trap>
+	{
+		let args = args.as_ref().iter().copied().map(Into::into);
+		if index < Self::function_count() {
+			return self.execute_function(index, args)
+				.map_err(Error::FunctionExecution)
+				.map_err(wasmi::Trap::from)
+				.map(|v| v.map(Into::into))
+		} else if index < SubstrateExternals::function_count() + Self::function_count() {
+			return SubstrateExternals::execute_function(index - Self::function_count(), args, self)
+				.map_err(Error::FunctionExecution)
+				.map_err(wasmi::Trap::from)
+				.map(|v| v.map(Into::into))
 		} else {
-			debug_trace!(
-				target: "wasm-trace",
-				"*** Setting storage:  {} -> {}   [k={}]",
-				primitives::hexdisplay::ascii_format(&key),
-				HexDisplay::from(&value),
-				HexDisplay::from(&key),
-			);
+			Err(
+				wasmi::Trap::from(
+					Error::from(format!("Could not find host function with index: {}", index))
+				)
+			)
 		}
-		this.ext.set_storage(key, value);
-		Ok(())
-	},
-	ext_set_child_storage(
-		storage_key_data: *const u8,
-		storage_key_len: u32,
-		key_data: *const u8,
-		key_len: u32,
-		value_data: *const u8,
-		value_len: u32
-	) => {
-		let storage_key = this.memory.get(storage_key_data, storage_key_len as usize)
-			.map_err(|_| "Invalid attempt to determine storage_key in ext_set_child_storage")?;
-		let key = this.memory.get(key_data, key_len as usize)
-			.map_err(|_| "Invalid attempt to determine key in ext_set_child_storage")?;
-		let value = this.memory.get(value_data, value_len as usize)
-			.map_err(|_| "Invalid attempt to determine value in ext_set_child_storage")?;
-		if let Some(_preimage) = this.hash_lookup.get(&key) {
-			debug_trace!(
-				target: "wasm-trace", "*** Setting child storage: {} -> %{} -> {}   [k={}]",
-				primitives::hexdisplay::ascii_format(&storage_key),
-				primitives::hexdisplay::ascii_format(&_preimage),
-				HexDisplay::from(&value),
-				HexDisplay::from(&key)
-			);
-		} else {
-			debug_trace!(
-				target: "wasm-trace", "*** Setting child storage: {} ->  {} -> {}   [k={}]",
-				primitives::hexdisplay::ascii_format(&storage_key),
-				primitives::hexdisplay::ascii_format(&key),
-				HexDisplay::from(&value),
-				HexDisplay::from(&key)
-			);
+	}
+}
+
+impl_wasm_host_interface! {
+	#[inherent_externals]
+	impl FunctionExecutor where this {
```

Comment:
I wanted to do the same, but it would have required a much more complicated declarative macro. I want to change the macro to a proc-macro anyway and then it will be better :)